tests/generic_config_updater/test_incremental_qos.py

```diff
@@ -25,69 +38,185 @@ def ensure_dut_readiness(duthost):
+    topo = tbinfo['topo']['name']
+
+    if "t1" in topo:
+        spine_router_count = 0
+        tor_router_count = 0
+        for neighbor in device_neighbor_metadata:
+            neighbor_data = device_neighbor_metadata[neighbor]
+            if neighbor_data['type'] == "SpineRouter":
+                spine_router_count += 1
+            elif neighbor_data['type'] == "ToRRouter":
+                tor_router_count += 1
+        return spine_router_count, tor_router_count
+
+    elif "t0" in topo:
+        leaf_router_count = 0
+        server_count = 0
+        for neighbor in device_neighbor_metadata:
+            neighbor_data = device_neighbor_metadata[neighbor]
+            if neighbor_data['type'] == "LeafRouter":
+                leaf_router_count += 1
+            elif neighbor_data['type'] == "Server":
+                server_count += 1
+        return leaf_router_count, server_count
+
+
+def get_neighbor_type_to_pg_headroom_map(duthost):
     """
-    Prepares config db by setting BUFFER_POOL key and field to specified value. If value is empty string or None, delete the current entry. 
+    Calculates pg headroom based on the present neighbor types
 
     Args:
         duthost: DUT host object
-        configdb_field: field in config_db BUFFER_POOL table of the form key/value
-        value: BUFFER_POOL table value to be set
+
+    Returns:
+        A map of neighbor type to its corresponding pg headroom value
     """
+    config_db_json = json.loads(duthost.shell("show runningconfig all")["stdout"])
+    device_neighbor_metadata = config_db_json["DEVICE_NEIGHBOR_METADATA"]
+    interfaces_data = config_db_json["PORT"]
+    neighbor_set = set()
+    neighbor_to_interface_map = {}
+    neighbor_to_type_map =  {}
+    neighbor_type_to_pg_headroom_map = {}
+
+    for neighbor in device_neighbor_metadata:
+        neighbor_set.add(neighbor)
+        neighbor_data = device_neighbor_metadata[neighbor]
+        neighbor_to_type_map[neighbor] = neighbor_data['type']
 
-    configdb_field_elements = configdb_field.split('/')
-    pytest_assert((len(configdb_field_elements) == 2), "Configdb field not identifiable")
+    for interface in interfaces_data:
+        for neighbor in neighbor_set:
+            if neighbor in json.dumps(interfaces_data[interface]):
+                neighbor_to_interface_map[neighbor] = interface
+                break
+
+    for neighbor in neighbor_set:
+        interface = neighbor_to_interface_map[neighbor]
+
+        cable_length = duthost.shell('sonic-db-cli CONFIG_DB hget "CABLE_LENGTH|AZURE" {}'.format(interface))['stdout']
+        port_speed = duthost.shell('sonic-db-cli CONFIG_DB hget "PORT|{}" speed'.format(interface))['stdout']
+
+        expected_profile = 'pg_lossless_{}_{}_profile'.format(port_speed, cable_length)
+
+        xoff = int(duthost.shell('redis-cli hget "BUFFER_PROFILE_TABLE:{}" xoff'.format(expected_profile))['stdout'])
```

Comment:
>redis-cli [](http://example.com/codeflow?start=34&length=9)

`redis-cli` will read AppDB (database 0) by default. I think you need `sonic-db-cli CONFIG_DB`.